scheduler/src/cook/mesos/util.clj

```diff
@@ -24,7 +24,56 @@
             [metatransaction.core :refer (db)]
             [metrics.timers :as timers]
             [plumbing.core :as pc :refer (map-vals map-keys)])
-  (:import [java.util Date]))
+  (:import
+    [com.google.common.cache Cache]
+    [com.google.common.cache CacheBuilder]
+    [java.util.concurrent TimeUnit]
+    [java.util Date]))
+
+
+(defn new-cache []
+  "Build a new cache"
+  (-> (CacheBuilder/newBuilder)
+      (.maximumSize 1000000)
+      ;; if its not been accessed in 2 hours, whatever is going on, its not being visted by the
+      ;; scheduler loop anymore. E.g., its probably failed/done and won't be needed. So,
+      ;; lets kick it out to keep cache small.
+      (.expireAfterAccess 2 TimeUnit/HOURS)
+      (.build)))
+
+(defn lookup-cache!
+  "Generic cache. Caches under a key (extracted from the item with extract-key-fn. Uses miss-fn to fill
+  any misses. Caches only postive hits where both functions return non-nil"
+  [^Cache cache extract-key-fn miss-fn item]
+  (let [key (extract-key-fn item)]
+    (if key
+      (if-let [result (.getIfPresent cache key)]
+        result ; we got a hit.
+        (let [new-result (miss-fn item)]
+          ; Only cache non-nil
+          (when new-result
+            (.put cache key new-result))
+          new-result))
+      (miss-fn item))))
+
+(defn lookup-cache-datomic-entity!
+  "Specialized function for caching where datomic entities are the key.
+  Extracts :db/id so that we don't keep the entity alive in the cache"
+  [cache miss-fn entity]
+  (lookup-cache! cache :db/id miss-fn entity))
+
+(defonce ^Cache job-ent->resources-cache (new-cache))
+(defonce ^Cache categorize-job-cache (new-cache))
+(defonce ^Cache task-ent->user-cache (new-cache))
+(defonce ^Cache task->feature-vector-cache (new-cache))
+
+(defn flush-caches!
+  "Flush the caches. Needed in unit tests. We centralize initialization by using it to initialize the caches too."
+  []
+  (.invalidateAll job-ent->resources-cache)
+  (.invalidateAll categorize-job-cache)
+  (.invalidateAll task-ent->user-cache)
+  (.invalidateAll task->feature-vector-cache))
```

Comment:
Is this called anywhere other than from unit tests? If not, can we move it to the unit tests?

Reply:
Done.

Reply:
It still seems to be here